source/common/http/conn_pool_grid.cc

```diff
@@ -232,9 +228,11 @@ absl::optional<ConnectivityGrid::PoolIterator> ConnectivityGrid::createNextPool(
   // HTTP/3 is hard-coded as higher priority, H2 as secondary.
   ConnectionPool::InstancePtr pool;
   if (pools_.empty()) {
-    pool = Http3::allocateConnPool(dispatcher_, random_generator_, host_, priority_, options_,
-                                   transport_socket_options_, state_, time_source_,
-                                   quic_stat_names_, scope_);
+    auto h3_pool = Http3::allocateConnPool(dispatcher_, random_generator_, host_, priority_,
+                                           options_, transport_socket_options_, state_,
+                                           time_source_, quic_stat_names_, scope_);
+    h3_pool->setConnectResultCallback(*this);
```

Comment:
Instead of using a setter here, how about adding an additional constructor argument which is an option ref to the connect callbacks?